plugins/isp/internode_usage

```diff
@@ -83,31 +83,25 @@ fi
 
 xpath_extract() {
 	xpath="$1"
-	# use trap to output an error message but let the caller get the status
-	# shellcheck disable=SC2064
-	trap "echo error extracting ${xpath} >&2" EXIT
 	xpath -q -n -e "${xpath}" \
-		| sed 's/<\([^>]*\)>\([^<]*\)<[^>]*>/\2/;s^N/A^U^'
-	trap - EXIT
+		| sed 's/<\([^>]*\)>\([^<]*\)<[^>]*>/\2/;s^N/A^U^' \
+		|| (echo "error extracting ${xpath}" >&2 \
+		    && false)
 }
 
 xpath_extract_attribute() {
 	xpath="$1"
-	# use trap to output an error message but let the caller get the status
-	# shellcheck disable=SC2064
-	trap "echo error extracting attribute at ${xpath} >&2" EXIT
-	xpath -q -n -e "${xpath}" \
-		| sed 's/.*="\([^"]\+\)".*/\1/'
-	trap - EXIT
+	(xpath -q -n -e "${xpath}" \
+		| sed 's/.*="\([^"]\+\)".*/\1/') \
+		|| (echo "error extracting attribute at ${xpath}" >&2 \
+		    && false)
 }
 
 fetch() {
-	# shellcheck disable=SC2064 disable=SC2154
-	# use trap to output an error message but let the caller get the status
-	trap "echo error fetching ${*} for user ${internode_api_login} >&2" EXIT
-	# shellcheck disable=SC2086 disable=SC2154
-	curl -u "${internode_api_login}:${internode_api_password}" -f ${CURL_ARGS} "$@"
-	trap - EXIT
+	# shellcheck disable=SC2154
+	curl -u "${internode_api_login}:${internode_api_password}" -f ${CURL_ARGS} "$@" \
+		|| (echo "error fetching ${*} for user ${internode_api_login}" >&2 \
```

Comment:
Just a minor nitpick: `{ ...; }` (instead of `(...)`) would be more efficient here, since it does not require a fork.

Reply:
Heh, TIL. I never realised I could use `{ ... ; }`.

The difference is notable
```
$ time sh -c 'for i in `seq 10000`; do true && (false); done'
sh -c 'for i in `seq 10000`; do true && (false); done'  2.66s user 2.47s system 104% cpu 4.886 total
$ time sh -c 'for i in `seq 10000`; do true && { false; }; done'
sh -c 'for i in `seq 10000`; do true && { false; }; done'  0.02s user 0.00s system 98% cpu 0.019 total
```

Reply:
Nice benchmark!

But to be fair: executing non-builtin statements (e.g. calling `curl`) is quite common in a script. Thus the difference in reality is probably not as noticeable as the above numbers may imply.